### PR TITLE
console: add support for UI plugin for the operator

### PR DESCRIPTION
**Pull request by weirdwiz**
- Updated the clusterversion_controller.go file to add a new method ensureConsolePlugin() responsible for creating/updating the ConsolePlugin CR, relevant Service, and the UI pod.

### CODE REVIEW

**Comment by weirdwiz** [comment]
PTAL @bipuladh

File: Makefile [hunk 1]
```diff
@@ -116,6 +117,7 @@ bundle: manifests kustomize operator-sdk yq ## Generate bundle manifests and met
 	rm -rf ./bundle
 	$(OPERATOR_SDK) generate kustomize manifests -q
 	cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
+    cd config/console && $(KUSTOMIZE) edit set image odf-client-console=$(ODF_CLIENT_CONSOLE_IMG)
```
Comment (Madhu-1) [inline Makefile:120]: Fix the indentation problem.

File: config/console/console_init.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: console
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: odf-client-console
```
Comment (Madhu-1) [inline config/console/console_init.yaml:8]: Please replace odf with ocs here, as there is no ODF operator here.

File: config/console/console_init.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+              subPath: nginx.conf
+            - name: odf-client-console-nginx-log
+              mountPath: /var/log/nginx
+            - name: odf-client-console-nginx-tmp
+              mountPath: /var/lib/nginx/tmp
+      tolerations:
+      - effect: NoSchedule
+        key: node.ocs.openshift.io/storage
+        operator: Equal
+        value: "true"
```
Comment (Madhu-1) [inline config/console/console_init.yaml:54]: we dont have any tolerations in client clusters.

File: config/console/kustomization.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+namePrefix: odf-
```
Comment (Madhu-1) [inline config/console/kustomization.yaml:1]: change odf to ocs

File: console/console.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	pluginBasePath = "/"
+
+	pluginDisplayName = "ODF Client Console"
+
+	servicePortName         = "console-port"
+	serviceSecretAnnotation = "service.alpha.openshift.io/serving-cert-secret-name"
+	serviceLabelKey         = "app.kubernetes.io/name"
+)
+
+func GetService(port int, namespace string) *apiv1.Service {
```
Comment (Madhu-1) [inline console/console.go:24]: take the input as int32 which avoids conversion below

File: console/console.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+				},
+			},
+			Selector: map[string]string{
+				serviceLabelKey: pluginName,
+			},
+		},
+	}
+}
+
+func GetDeployment(namespace string) *appsv1.Deployment {
```
Comment (Madhu-1) [inline console/console.go:52]: its not returning any template here why do we need a separate function for it?

File: console/console.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	return &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pluginName,
+			Namespace: namespace,
+		},
+	}
+}
+
+func GetConsolePluginCR(consolePort int, serviceNamespace string) *consolev1alpha1.ConsolePlugin {
```
Comment (Madhu-1) [inline console/console.go:61]: it should be called `GetConsolePlugin`?

File: console/console.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	return &consolev1alpha1.ConsolePlugin{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: pluginName,
```
Comment (Madhu-1) [inline console/console.go:64]: no namespace is required here?

File: console/nginx_conf.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
+package console
+
+var NginxConf = `
```
Comment (Madhu-1) [inline console/nginx_conf.go:3]: do we need this to be exported?

File: controllers/clusterversion_controller.go [hunk 10]
```diff
@@ -99,6 +103,10 @@ func (c *ClusterVersionReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		},
 	)
 
+	if err := c.ensureConsolePlugin(); err != nil {
```
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:106]: why do we need to call it here?

File: controllers/clusterversion_controller.go [hunk 11]
```diff
@@ -356,3 +369,49 @@ func (c *ClusterVersionReconciler) getOperatorConfig() (*corev1.ConfigMap, error
...
+	nginxConf := console.GetNginxConfiguration()
+
+	consoleDeployment := console.GetDeployment(c.OperatorNamespace)
+	err := c.Client.Get(context.TODO(), types.NamespacedName{
+		Name:      consoleDeployment.Name,
+		Namespace: consoleDeployment.Namespace,
+	}, consoleDeployment)
+	if err != nil {
+		return err
+	}
```
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:383]: Why do we need to do Get here?

File: controllers/clusterversion_controller.go [hunk 12]
```diff
@@ -356,3 +369,49 @@ func (c *ClusterVersionReconciler) getOperatorConfig() (*corev1.ConfigMap, error
...
+
+	consoleConfigMap := console.GetNginxConfConfigMap(c.OperatorNamespace)
+	_, err = controllerutil.CreateOrUpdate(context.TODO(), c.Client, consoleConfigMap, func() error {
```
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:386]: use c.ctx instead of TODO

File: controllers/clusterversion_controller.go [hunk 13]
```diff
@@ -356,3 +369,49 @@ func (c *ClusterVersionReconciler) getOperatorConfig() (*corev1.ConfigMap, error
...
+	})
+
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	// Create/Update ODF console Service
+	consoleService := console.GetService(c.ConsolePort, c.OperatorNamespace)
+	_, err = controllerutil.CreateOrUpdate(context.TODO(), c.Client, consoleService, func() error {
+		return controllerutil.SetControllerReference(consoleDeployment, consoleService, c.Scheme)
```
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:401]: you might need to do update inside the function to replace the spec, please check

File: controllers/clusterversion_controller.go [hunk 14]
```diff
@@ -356,3 +369,49 @@ func (c *ClusterVersionReconciler) getOperatorConfig() (*corev1.ConfigMap, error
...
+	})
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	// Create/Update ODF console ConsolePlugin
+	consolePlugin := console.GetConsolePluginCR(c.ConsolePort, c.OperatorNamespace)
+	_, err = controllerutil.CreateOrUpdate(context.TODO(), c.Client, consolePlugin, func() error {
+		return nil
```
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:410]: you might need to do update inside the function to replace the spec, please check

File: hack/make-bundle-vars.mk [hunk 15]
```diff
@@ -52,6 +52,11 @@ CSI_ADDONS_BUNDLE_IMAGE_NAME ?= k8s-bundle
 CSI_ADDONS_BUNDLE_IMAGE_TAG ?= v0.5.0
 CATALOG_IMAGE_NAME ?= $(IMAGE_NAME)-catalog
 
+ODF_CLIENT_CONSOLE_IMG_NAME ?= odf-client-console
```
Comment (Madhu-1) [inline hack/make-bundle-vars.mk:55]: ocs-client-console

File: hack/make-bundle-vars.mk [hunk 16]
```diff
@@ -52,6 +52,11 @@ CSI_ADDONS_BUNDLE_IMAGE_NAME ?= k8s-bundle
...
+ODF_CLIENT_CONSOLE_IMG_TAG ?= latest
+ODF_CLIENT_CONSOLE_IMG_LOCATION ?= quay.io/dkamboj
```
Comment (Madhu-1) [inline hack/make-bundle-vars.mk:57]: change the location

**Comment by Madhu-1** [comment]
@weirdwiz please take a look at CI failures

**Comment by weirdwiz** [comment]
updates: refactored a couple of functions and addressed comments

PTAL @Madhu-1 @jarrpa @nb-ohad

File: bundle/manifests/ocs-client-operator.clusterserviceversion.yaml [hunk 17]
```diff
@@ -668,6 +669,134 @@ spec:
...
+              app.kubernetes.io/name: ocs-client-console
+          strategy: {}
+          template:
+            metadata:
+              labels:
+                app: odf-client-console
+                app.kubernetes.io/name: ocs-client-console
+            spec:
+              containers:
+              - image: /:latest
```
Comment (Madhu-1) [inline bundle/manifests/ocs-client-operator.clusterserviceversion.yaml:688]: image name is missing here

File: bundle/manifests/ocs-client-operator.clusterserviceversion.yaml [hunk 18]
```diff
@@ -668,6 +669,134 @@ spec:
               - configMap:
                   name: ocs-client-operator-csi-images
                 name: csi-images
+      - label:
+          app: odf-client-console
```
Comment (Madhu-1) [inline bundle/manifests/ocs-client-operator.clusterserviceversion.yaml:673]: ocs-client-console

File: bundle/manifests/ocs-client-operator.clusterserviceversion.yaml [hunk 19]
```diff
@@ -668,6 +669,134 @@ spec:
...
+                  secretName: ocs-client-console-serving-cert
+              - configMap:
+                  name: ocs-client-console-nginx-conf
+                name: ocs-client-console-nginx-conf
+              - emptyDir: {}
+                name: ocs-client-console-nginx-log
+              - emptyDir: {}
+                name: ocs-client-console-nginx-tmp
+      - label:
+          app: odf-client-console
```
Comment (Madhu-1) [inline bundle/manifests/ocs-client-operator.clusterserviceversion.yaml:737]: ocs-client-console

File: controllers/clusterversion_controller.go (hunk 10 above)
Comment (Madhu-1) [reply controllers/clusterversion_controller.go:106]: still its pending. why we need it here

File: controllers/clusterversion_controller.go [hunk 20]
```diff
@@ -356,3 +364,68 @@ func (c *ClusterVersionReconciler) getOperatorConfig() (*corev1.ConfigMap, error
...
+			Name:      console.NginxConfigMapName,
+			Namespace: c.OperatorNamespace,
+		},
+		Data: map[string]string{
+			"nginx.conf": nginxConf,
+		},
+	}
+	err := c.createOrUpdate(nginxConfigMap, func() error {
+		if consoleConfigMapData := nginxConfigMap.Data["nginx.conf"]; consoleConfigMapData != nginxConf {
+			c.log.Info(fmt.Sprintf("set the ConfigMap %s data as '%s'", nginxConfigMap.Name, nginxConf))
```
Comment (Madhu-1) [inline controllers/clusterversion_controller.go:389]: can you remove this logging?

Comment (weirdwiz) [reply controllers/clusterversion_controller.go]: removed!

File: controllers/clusterversion_controller.go (hunk 11 above)
Comment (weirdwiz) [reply controllers/clusterversion_controller.go]: we need to get the UID of the deployment of the console to set the owner of the configmaps and service

odf-console has the console deployment as the owner of these resources. not sure if setting the client-operator as owner is going to have a detrimental affect

File: controllers/clusterversion_controller.go (hunk 10 above)
Comment (weirdwiz) [reply controllers/clusterversion_controller.go]: removed, the change wasn't staged

**Comment by weirdwiz** [comment]
updated, PTAL @Madhu-1 :D

**Review by Madhu-1 — COMMENTED** [review]
@weirdwiz can you make sure the latest changes works fine without any problem, if yes am good to get it merged.

**Pull request merged**